REST_API_COMPATIBILITY.md

```diff
@@ -0,0 +1,182 @@
+
+In places where `ToXContent#toXContent` is not used for serialization, then the requested compatibility can also be found in the `RestRequest` object and used accordingly. It is generally discouraged to push the requested compatibility version down through the transport actions to allow keeping the REST compatibility a REST layer concern.
+
+### De-serializing
+
+Accepting a payload from a prior version's REST request is a bit more difficult sending a compatible response. There are a couple different approaches for de-serialization, however, in general all of them use a request specific parser.  `XContentParser` has a `getRestApiVersion()` method that could be used in manner that is near identical to serialization. However, in practice most parsers are invoked via a `ConstructingObjectParser` or an `ObjectParser` and quite often they are daisy chained together via the `NamedXContentRegistry`.
+
+Both `ConstructingObjectParser` and `ObjectParser` use `ParseField`'s to statically declare the relationship between a named key and value pairing. A `ParseField` can be made version aware, meaning that it will match on the incoming payload if the requested compatibility is the correct version.
+
+For example:
+
+```java
+PARSER.declareInt(MyPojo::setMax,
+       new ParseField("maximum", "limit").forRestApiVersion(RestApiVersion.equalTo(RestApiVersion.V_7)));
+PARSER.declareInt(MyPojo::setMax,
+       new ParseField("maximum").forRestApiVersion(RestApiVersion.onOrAfter(RestApiVersion.V_8)));
+```
+
+The above example is for code that live in the v8 branch of code. It reads the `maximum` value from the request for both v7 and v8. However, if compatibility is requested it will also allow `limit` in the payload.  If `limit` is used a warning will be emitted.
+
+The version in `forRestApiVersion` is reference to when the declaration is valid. Assuming v8 is the master branch and all changes start in the master branch then get back ported. The above text is what would be applicable for the v8 branch of code. The first line of code is essentially ignored except for when compatibility with v7 is requested. When back-porting this change to the 7.x branch, the first line would be identical, and the second line would be omitted.
+
+The above strategy works well for single fields, but could get overly complex very fast for large multiple field changes. For more complex de-serialization changes there is also support to construct a `NamedXContentRegistry` with some "normal" entries as well some entries that are only applied when compatibility with the prior version is requested. Additionally there is a pseudo standard method `fromXContent` which also has access to the `XContentParser` for that request which can be used to conditionally change how to parse the input.
```

Comment:
maybe we could provide examples for named xcontent registry and `fromXContent` ?